README.md


# Exometer Collectd reporter #

Copyright (c) 2014 Basho Technologies, Inc.  All Rights Reserved.

__Version:__ Nov 6 2018 08:44:40

__Authors:__ Ulf Wiger ([`ulf.wiger@feuerlabs.com`](mailto:ulf.wiger@feuerlabs.com)), Magnus Feuer ([`magnus.feuer@feuerlabs.com`](mailto:magnus.feuer@feuerlabs.com)).

[![Travis][travis badge]][travis]
[![Hex.pm Version][hex version badge]][hex]
[![Hex.pm License][hex license badge]][hex]
[![Erlang Versions][erlang version badge]][travis]
[![Build Tool][build tool]][hex]


#### <a name="exometer_report_collectd">exometer_report_collectd</a> ####

The collectd reporter communicates with a local `collectd` process
through its unix socket protocol. All subscribed-to metric-datapoint
values received by the reporter are immediately forwarded to
`collectd`. Once a value has been forwarded, the reporter continuously
refreshes the value toward `collectd` at a configurable interval in order
to keep it from expiring inside `collectd`.

If the `collectd` connection is lost, the reporter will attempt to reconnect to it
at a configurable interval.

All metrics reported to collectd will be have identifiers formatted as follows:

```
HostName/PluginName-PluginInstance/Type-Metric_DataPoint
```

+ `HostName`<br />Host name of the entry.<br />Configurable through the `hostname` application environment parameter.<br />Default is the value returned by `netadm:localhost()`.

+ `PluginName`<br />The collectd plugin name.<br />Configurable through the `plugin_name` application environment parameter.<br />Default is `exometer`.

+ `PluginInstance`<br />The instance ID to use for the plugin.<br />Configurable through the `plugin_instance` application environment parameter.<br />Default is the erlang node name in the left hand side of the value
    returned by `node()`.

+ `Type`<br />Type assigned to the reported value.<br />The type is looked up through the `type_map`.<br />The given metric and data points are used as a key in a list format,
    such as `[ db, cache, hits, median ]`. The type that is resolved from
    the metric/data point will be used as the `Type` component in the
    collectd identifier. Please see types.db(5) for a list of available
    collectd types.<br />Default for `Type` is 'gauge'.

+ `Metric`<br />The name of the metric. The atoms in the metric list will be converted
    to a string separated by `_`. Thus `[ db, cache, hits ]` will be converted
    to `db_cache_hits`.

+ `DataPoint`<br />The data point of the given metric.
Will be added to the end of the metrics string.

Please see [Configuring collectd reporter](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_collectd_reporter) for details on the
application environment parameters listed above.


#### <a name="Setting_up_subscriptions">Setting up subscriptions</a> ####

A subscription can either be statically configured, or dynamically
setup from within the code using Exometer. For details on statically
configured subscriptions, please see [Configuring static subscriptions](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_static_subscriptions).

A dynamic subscription can be setup with the following call:

```erlang

exometer_report:subscribe(Recipient, Metric, DataPoint, Inteval)
```

`Recipient` is the name of a reporter.


#### <a name="Configuring_type_-_entry_maps">Configuring type - entry maps</a> ####

The dynamic method of configuring defaults for `exometer` entries is:

```erlang

exometer_admin:set_default(NamePattern, Type, Default)
```

Where `NamePattern` is a list of terms describing what is essentially
a name prefix with optional wildcards (`'_'`). A pattern that
matches any legal name is `['_']`.

`Type` is an atom defining a type of metric. The types already known to
`exometer`, `counter`, `fast_counter`, `ticker`, `uniform`, `histogram`,
`spiral`, `netlink`, and `probe` may be redefined, but other types can be
described as well.

`Default` is either an `#exometer_entry{}` record (unlikely), or a list of
`{Key, Value}` options, where the keys correspond to `#exometer_entry` record
attribute names. The following attributes make sense to preset:

```erlang

{module, atom()}              % the callback module
{status, enabled | disabled}  % operational status of the entry
{cache, non_neg_integer()}    % cache lifetime (ms)
{options, [{atom(), any()}]}  % entry-specific options
```

Below is an example, from `exometer/priv/app.config`:

```erlang

{exometer, [
    {defaults, [
        {['_'], function , [{module, exometer_function}]},
        {['_'], counter  , [{module, exometer}]},
        {['_'], histogram, [{module, exometer_histogram}]},
        {['_'], spiral   , [{module, exometer_spiral}]},
        {['_'], duration , [{module, exometer_folsom}]},
        {['_'], meter    , [{module, exometer_folsom}]},
        {['_'], gauge    , [{module, exometer_folsom}]}
    ]}
]}
```

In systems that use CuttleFish, the file `exometer/priv/exometer.schema`
contains a schema for default settings. The setup corresponding to the above
defaults would be as follows:

```ini

exometer.template.function.module  = exometer_function
exometer.template.counter.module   = exometer
exometer.template.histogram.module = exometer_histogram
exometer.template.spiral.module    = exometer_spiral
exometer.template.duration.module  = exometer_folsom
exometer.template.meter.module     = exometer_folsom
exometer.template.gauge.module     = exometer_folsom
```


#### <a name="Configuring_static_subscriptions">Configuring static subscriptions</a> ####

Static subscriptions, which are automatically setup at exometer
startup without having to invoke `exometer_report:subscribe()`, are
configured through the report sub section under exometer.

Below is an example, from `exometer/priv/app.config`:

```erlang

{exometer, [
    {report, [
        {subscribers, [
            {exometer_report_collectd, [db, cache, hits], mean, 2000, true},
            {exometer_report_collectd, [db, cache, hits], max, 5000, false}
        ]}
    ]}
]}
```

The `report` section configures static subscriptions and reporter
plugins. See [Configuring reporter plugins](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_reporter_plugins) for details on
how to configure individual plugins.

The `subscribers` sub-section contains all static subscriptions to be
setup att exometer applications start. Each tuple in the prop list
should be of one of the following formats:

* `{Reporter, Metric, DataPoint, Interval}`

* `{Reporter, Metric, DataPoint, Interval, RetryFailedMetrics}`

* `{Reporter, Metric, DataPoint, Interval, RetryFailedMetrics, Extra}`

* `{apply, {M, F, A}}`

* `{select, {MatchPattern, DataPoint, Interval [, Retry [, Extra] ]}}`

In the case of `{apply, M, F, A}`, the result of `apply(M, F, A)` must
be a list of `subscribers` tuples.

In the case of `{select, Expr}`, a list of metrics is fetched using
`exometer:select(MatchPattern)`, where the result must be on the form
`{Key, Type, Status}` (i.e. what corresponds to `'$_'`).
The rest of the items will be applied to each of the matching entries.

The meaning of the above tuple elements is:

+ `Reporter :: module()`<br />Specifies the reporter plugin module, such as`exometer_report_collectd` that is to receive updated metric's data
points.

+ `Metric :: [atoms()]`<br />Specifies the path to a metric previously created with an`exometer:new()` call.

+ `DataPoint` ::  atom() | [atom()]'<br />Specifies the data point within the given metric to send to the
    receiver. The data point must match one of the data points returned by`exometer:info(Name, datapoints)` for the given metrics name.

+ `Interval` :: integer()' (milliseconds)<br />Specifies the interval, in milliseconds, between each update of the
given metric's data point. At the given interval, the data point will
be samples, and the result will be sent to the receiver.

+ `RetryFailedMetrics :: boolean()`<br />Specifies if the metric should be continued to be reported
    even if it is not found during a reporting cycle. This would be
    the case if a metric is not created by the time it is reported for
    the first time. If the metric will be created at a later time,
    this value should be set to true. Set this value to false if all
    attempts to report the metric should stop if when is not found.
    The default value is `true`.

+ `Extra :: any()`<br />Provides a means to pass along extra information for a given
   subscription. An example is the `syntax` option for the SNMP reporter,
   in which case `Extra` needs to be a property list.

Example configuration in sys.config, using the `{select, Expr}` pattern:

```erlang

[
 {exometer, [
             {predefined,
              [{[a,1], counter, []},
               {[a,2], counter, []},
               {[b,1], counter, []},
               {[c,1], counter, []}]},
             {report,
              [
               {reporters,
                [{exometer_report_tty, []}]},
               {subscribers,
                [{select, {[{ {[a,'_'],'_','_'}, [], ['$_']}],
                           exometer_report_tty, value, 1000}}]}
              ]}
            ]}
].

```

This will activate a subscription on `[a,1]` and `[a,2]` in the
`exometer_report_tty` reporter, firing once per second.


#### <a name="Configuring_collectd_reporter">Configuring collectd reporter</a> ####

Below is an example of the collectd reporter application environment, with
its correct location in the hierarchy:

```erlang

{exometer, [
    {report, [
        {reporters, [
            {exometer_report_collectd, [
                {reconnect_interval, 10},
                {refresh_interval, 20},
                {read_timeout, 5000},
                {connect_timeout, 8000},
                {hostname, "testhost"},
                {path, "/var/run/collectd-unixsock"},
                {plugin_name, "testname"},
                {plugin_instance, "testnode"},
                {type_map,
                    [{[db, cache, hits, max], "gauge"}]
                }
            ]}
        ]}
    ]}
]}
```

The following attributes are available for configuration:

+ `reconnect_interval` (seconds - default: 30)<br />Specifies the duration between each reconnect attempt to a collectd
server that is not available. Should the server either be unavailable
at exometer startup, or become unavailable during exometer's
operation, exometer will attempt to reconnect at the given number of
seconds.

+ `refresh_interval` (seconds - default: 10)<br />Specifies how often a value, which has not been updated by exometer,
is to be resent with its current value to collectd. If collectd does
not see an identifier updated at a given number of seconds (see
Timeout in collectd.conf(5)), it will be removed from the list of
maintained identifiers.

+ `read_timeout` (milliseconds - default: 5000)<br />Specifies how long the collectd reporter plugin shall wait for an
acknowledgement from collectd after sending it an updated value.  If
an acknowledgment is not received within the given interval, the
plugin will disconnect from the collectd server and reconnect to it
after the given reconnect interval (see item one above).

+ `connect_timeout` (milliseconds - default: 5000)<br />Specifies how long the collectd reporter plugin shall wait for a unix
socket connection to complete before timing out. A timed out
connection attempt will be retried after the reconnect interval has
passed see item 1 above).

+ `path` (file path - default: "/var/run/collectd-unixsock")<br />Specifies the path to the named unix socket that collectd is listening
on. When exometer starts and loads the collectd reporter plugin, the
plugin will connect to the given socket.

+ `plugin_name` (string - default: "exometer")<br />Specifies the plugin name to use when constructing an collectd identifier.
    Please see [Configuring collectd reporter](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_collectd_reporter) for details.

+ `plugin_instance` (string - default: left hand side of `node()`)<br />Specifies the plugin instance id to use when constructing an collectd identifier.
    Please see [Configuring collectd reporter](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_collectd_reporter) for details.

+ `plugin_instance` (string - default: left hand side of `node()`)<br />Specifies the plugin instance id to use when constructing an collectd identifier.
    Please see [Configuring collectd reporter](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_collectd_reporter) for details.

+ `hostname` (string - default: `net_adm:localhost()`)<br />Specifies the host name to use when constructing an collectd identifier.
    Please see [Configuring collectd reporter](https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/README.md#Configuring_collectd_reporter) for details.

+ `type_map` (prop list - default: n/a)<br />Specifies the mapping between metrics/datapoints and the collectd type
to use when sending an updated metric value. See below.

Type maps must be provided since there is no natural connection
between the type of a metric/datapoint pair and an identifier in
collectd. The `type_map` consists of a prop list that converts a path
to a metric/datapoint to a string that can be used as a type when
reporting to collectd.

The key part of each element in the list consists of a list of atoms
that matches the name of the metrics, with the name of the data point
added as a final element. If the metric is identified as `[ webserver,
https, get_count ]`, and the data point is called `total`, the key in
the type_map would be `[ webserver, https, get_count, total ]`, The
value part of a property is the type string to use when reporting to
collectd. Please see types.db(5) for a list of available collectd
types.  A complete entry in the `type_map` list would be: `{ [
webserver, https, get_count, total ], "counter" }`.

[travis]: https://travis-ci.org/Feuerlabs/exometer_collectd
[travis badge]: https://img.shields.io/travis/Feuerlabs/exometer_collectd/master.svg?style=flat-square
[hex]: https://hex.pm/packages/exometer_collectd
[hex version badge]: https://img.shields.io/hexpm/v/exometer_collectd.svg?style=flat-square
[hex license badge]: https://img.shields.io/hexpm/l/exometer_collectd.svg?style=flat-square
[erlang version badge]: https://img.shields.io/badge/erlang-18--21-blue.svg?style=flat-square
[build tool]: https://img.shields.io/badge/build%20tool-rebar3-orange.svg?style=flat-square


## Modules ##


<table width="100%" border="0" summary="list of modules">
<tr><td><a href="https://github.com/Feuerlabs/exometer_collectd/blob/master/doc/exometer_report_collectd.md" class="module">exometer_report_collectd</a></td></tr></table>



CHANGELOG.md
# Changelog

## [v1.1.0](https://github.com/Feuerlabs/exometer_collectd/tree/v1.1.0) (2018-11-06)
[Full Changelog](https://github.com/Feuerlabs/exometer_collectd/compare/1.0.1...v1.1.0)

- Upgrade to rebar3
- Remove obsolete build configuration
- Update application metadata
- Upgrade dependency exometer_core


src/exometer_report_collectd.erl
%% -------------------------------------------------------------------
%%
%% Copyright (c) 2014 Basho Technologies, Inc.  All Rights Reserved.
%%
%%   This Source Code Form is subject to the terms of the Mozilla Public
%%   License, v. 2.0. If a copy of the MPL was not distributed with this
%%   file, You can obtain one at http://mozilla.org/MPL/2.0/.
%%
%% -------------------------------------------------------------------

%% @doc Custom reporting probe for Hosted Graphite.
%%
%% Collectd unix socket integration.
%% All data subscribed to by the plugin (through exosense_report:subscribe())
%% will be reported to collectd.
%%
%% Options:
%%
%% `{path, string()}' - The AFUNIX socket path to the collectd socket.
%% Default: `"/var/run/collectd-unixsock"'.
%%
%% `{connect_timeout, non_neg_integer()}' - Timeout, in milliseconds, for the
%% connect operation. Default: `5000' (ms).
%%
%% `{read_timeout, non_neg_integer()}' - Read timeout, in milliseconds, when
%% receiving replies on the AFUNIX socket. Default: `5000' (ms).
%%
%% `{reconnect_interval, non_neg_integer()}' - Time, in seconds, before
%% attempting to reconnect. Default: `30' (sec)
%%
%% `{connect_retries, non_neg_integer() | infinity}' - How many times to
%% try reconnecting before automatically disabling the reporter.
%% Default: `infinity'
%%
%% `{refresh_interval, non_neg_integer()}' - Time, in seconds, before
%% re-sending a metric value, if it hasn't been updated before then.
%% Default: `10' (sec).
%% @end

%% Collectd expects periodical metrics "refreshes", even if the values
%% have not changed. We do this through erlang:send_after() calls with
%% the metrics / value update to emit.
%%
%% Please note that exometer_report_collectd is still also a
%% exometer_report implementation.

-module(exometer_report_collectd).
-behaviour(exometer_report).

-export(
   [
    exometer_init/1,
    exometer_info/2,
    exometer_cast/2,
    exometer_call/3,
    exometer_report/5,
    exometer_subscribe/5,
    exometer_unsubscribe/4,
    exometer_newentry/2,
    exometer_setopts/4,
    exometer_terminate/2
   ]).


-define(SERVER, ?MODULE).

-include_lib("hut/include/hut.hrl").
-include_lib("exometer_core/include/exometer.hrl").

-define(CONNECT_TIMEOUT, 5000).
-define(CONNECT_RETRIES, infinity).
-define(RECONNECT_INTERVAL, 30). %% seconds
-define(READ_TIMEOUT, 5000).
-define(REFRESH_INTERVAL, 10). %% seconds
-define(DEFAULT_PATH, "/var/run/collectd-unixsock").

-record(st, {
          hostname = undefined,
          socket_path = undefined,
          plugin_name = undefined,
          plugin_instance = undefined,
          refresh_interval = ?REFRESH_INTERVAL,
          type_map = undefined,
          read_timeout = ?READ_TIMEOUT,
          connect_timeout = ?CONNECT_TIMEOUT,
	  connect_retries = infinity,
          reconnect_interval = ?RECONNECT_INTERVAL,
          socket = undefined}).

%% calendar:datetime_to_gregorian_seconds({{1970,1,1},{0,0,0}}).
-define(UNIX_EPOCH, 62167219200).

exometer_init(Opts) ->
    ?log(info, "Exometer(~p): Starting~n", [Opts]),
    SockPath = get_opt(path, Opts, ?DEFAULT_PATH),
    ConnectTimeout = get_opt(connect_timeout, Opts, ?CONNECT_TIMEOUT),
    ReconnectInterval =
        get_opt(reconnect_interval, Opts, ?RECONNECT_INTERVAL) * 1000,
    ConnectRetries = reconnect_init(
		       get_opt(connect_retries, Opts, ?CONNECT_RETRIES)),

    %% [ { metric, type }, ... ]
    ets:new(exometer_collectd, [named_table, {keypos, 1}, public, set]),

    %% Try to connect to collectd.
    St0 = #st{socket_path = SockPath,
	      reconnect_interval = ReconnectInterval,
	      hostname = check_hostname(
			   get_opt(hostname, Opts, "auto")),
	      plugin_name = get_opt(plugin_name, Opts, "exometer"),
	      plugin_instance = check_instance(
				  get_opt(plugin_instance, Opts, "auto")),
	      read_timeout = get_opt(read_timeout, Opts, ?READ_TIMEOUT),
	      connect_timeout = ConnectTimeout,
	      connect_retries = ConnectRetries,
	      refresh_interval = get_opt(refresh_interval, Opts,
					 ?REFRESH_INTERVAL) * 1000,
	      type_map = get_opt(type_map, Opts, undefined)
	     },
    case connect_collectd(SockPath, ConnectTimeout) of
        {ok, Sock} ->
            ?log(info, "Exometer collectd connection succeeded"),
	    {ok, St0#st{socket = Sock}};
        {error, _} = Error ->
            ?log(warning, "Exometer collectd connection failed; ~p. Retry in ~p~n",
		     [Error, ReconnectInterval]),
	    prepare_reconnect(),
	    {ok, St0}
    end.

reconnect_init(infinity) ->
    infinity;
reconnect_init(Retries) when is_integer(Retries), Retries >= 0 ->
    {1, Retries}.

reconnect_incr(infinity) ->
    {true, infinity};
reconnect_incr({N, Max}) ->
    N1 = N+1,
    {N < Max, {N1, Max}}.

exometer_subscribe(_Metric, _DataPoint, _Extra, _Interval, St) ->
    {ok, St}.

exometer_unsubscribe(Metric, DataPoint, _Extra, St) ->
    %% Kill off any refresh timers that we may have handle_log(info,  {
    %% refresh_metric, ...) will verify that the ets table has a key
    %% before it refreshes the metric in collectd and reschedules the
    %% next refresh operation.
    case ets:lookup(exometer_collectd, ets_key(Metric, DataPoint)) of
        [] -> ok;
        [{_, TRef}] ->
            ?log(info, "Canceling old timer through unsubscribe~n"),
            ets:delete(exometer_collectd, ets_key(Metric, DataPoint)),
            erlang:cancel_timer(TRef)
    end,
    {ok, St}.



%% Exometer report when no collectd socket connection exists.
exometer_report(_Metric, _DataPoint, _Extra, _Value, St)
  when St#st.socket =:= undefined ->
    {ok, St};

%% Invoked through the remote_exometer() function to
%% send out an update.
exometer_report(Metric, DataPoint, Extra, Value, St)  ->
    ?log(debug, "Report metric ~p_~p = ~p~n", [Metric, DataPoint, Value]),
    %% Cancel and delete any refresh timer, if it exists
    case ets:lookup(exometer_collectd, ets_key(Metric, DataPoint)) of
        [] -> ok;
        [{_, TRef}] ->
            %% We don't need to delete the old ets entry
            %% since it will be replaced by ets:insert()
            %% in report_exometer_()
            ?log(debug, "Canceling old timer~n"),
            erlang:cancel_timer(TRef)

    end,

    %% Report the value and setup a new refresh timer.
    {ok, report_exometer_(Metric, DataPoint, Extra, Value, St)}.

exometer_call(Unknown, From, St) ->
    ?log(info, "Unknown call ~p from ~p", [Unknown, From]),
    {ok, St}.

exometer_cast(Unknown, St) ->
    ?log(info, "Unknown cast: ~p", [Unknown]),
    {ok, St}.

exometer_info({exometer_callback, prepare_reconnect},
	      #st{reconnect_interval = Int,
		  connect_retries = Retries} = St) ->
    case reconnect_incr(Retries) of
	{true, Retries1} ->
	    reconnect_after(Int),
	    {ok, St#st{connect_retries = Retries1}};
	{false, _} ->
	    ?log(warning, "Reached max connection retries, disabling~n", []),
	    exometer_report:disable_me(?MODULE, St)
    end;
exometer_info({exometer_callback, refresh_metric,
               Metric, DataPoint, Extra, Value}, St) ->
    %% Make sure that we still have an entry in the ets table.
    %% If not, exometer_unsubscribe() has been called to remove
    %% the entry, and we should do nothing.
    case ets:lookup(exometer_collectd, ets_key(Metric, DataPoint)) of
        [] ->
             {ok, St};
        [{_, _TRef}] ->
            {ok, report_exometer_(Metric, DataPoint, Extra, Value, St)}
    end;
exometer_info({exometer_callback, reconnect}, St) ->
    case connect_collectd(St) of
        {ok, NSt} ->
            {ok, NSt};
        Err  ->
            Err,
	    prepare_reconnect(),
            {ok, St}
    end;
exometer_info(Unknown, St) ->
    ?log(info, "Unknown: ~p~n", [Unknown]),
    {ok, St}.

exometer_newentry(_Entry, St) ->
    {ok, St}.

exometer_setopts(_Metric, _Options, _Status, St) ->
    {ok, St}.

exometer_terminate(_, _) ->
    ignore.

report_exometer_(Metric, DataPoint, Extra, Value,
                 #st{hostname = HostName,
                     plugin_name = PluginName,
                     plugin_instance = PluginInstance,
                     socket = Sock,
                     type_map = TypeMap} = St) ->
    case get_type(TypeMap, Extra, ets_key(Metric, DataPoint)) of
        error ->
            ?log(warning,
               "Could not resolve ~p to a collectd type."
               "Update exometer_report_collectd -> type_map in app.config. "
               "Value lost~n", [ets_key(Metric, DataPoint)]),
            St;
        {ok, Type} ->
            Request = ["PUTVAL ", HostName, "/",
                       PluginName, "-", PluginInstance, "/",
                       Type, "-", name(Metric, DataPoint), $\s,
                       timestamp(), ":", value(Value), $\n],
            send_request(Sock, Request, Metric, DataPoint, Extra, Value, St)
    end.

send_request(Sock, Request, Metric, DataPoint, Extra, Value,
             #st{read_timeout = TOut} = St) ->
    try gen_tcp:send(Sock, Request) of
        ok ->
            case gen_tcp:recv(Sock, 0, TOut) of
                {ok, Bin} ->
                    %% Parse the reply
                    case parse_reply(Request, Bin, St) of
                        %% Reply is ok.
                        %% Ensure that we have periodical refreshes
                        %% of this value.
                        {ok, St1} ->
                            ?log(debug, "Setting up refresh~n"),
                            setup_refresh(
                              St1#st.refresh_interval, Metric,
                              DataPoint, Extra, Value),
                            St1;
                        %% Something went wrong with reply.
                        %% Do not refresh
                        _ -> St
                    end;
                _ ->
                    %% We failed to receive data,
                    %% close and setup later reconnect
                    ?log(warning, "Failed to receive. Will reconnect in ~p~n",
                             [St#st.reconnect_interval]),
                    maybe_reconnect_after(Sock),
                    St#st{socket = undefined}
            end;
        {error, enotconn} ->
            %% Socket is not connected, setup later reconnect
            ?log(warning, "Failed to send. Will reconnect in ~p~n",
                     [St#st.reconnect_interval]),
            maybe_reconnect_after(Sock),
            St#st{socket = undefined};
        _ ->
            St
    catch
        error:_ ->
            %% We failed to receive data, close and setup later reconnect
            ?log(warning, "Failed to send. Will reconnect in ~p~n",
                     [St#st.reconnect_interval]),
            maybe_reconnect_after(Sock),
            St#st {socket = undefined}
    end.

ets_key(Metric, DataPoint) ->
    Metric ++ [ DataPoint ].

%% Add metric and datapoint within metric
name(Metric, DataPoint) when is_integer(DataPoint) or is_float(DataPoint) ->
    name(Metric, value(DataPoint));

name(Metric, DataPoint) when is_atom(DataPoint) ->
    name(Metric, atom_to_list(DataPoint));

name(Metric, DataPoint) ->
    metric_to_string(Metric) ++ "_" ++ DataPoint.

metric_to_string([Final]) ->
    metric_elem_to_list(Final);

metric_to_string([H | T]) ->
    metric_elem_to_list(H) ++ "_" ++ metric_to_string(T).

metric_elem_to_list(E) when is_atom(E) ->
    atom_to_list(E);

metric_elem_to_list(E) when is_list(E) ->
    E;

metric_elem_to_list(E) when is_integer(E) ->
    integer_to_list(E).



%% Add value, int or float, converted to list
value(V) when is_integer(V) -> integer_to_list(V);
value(V) when is_float(V)   -> io_lib:format("~f", [V]);
value(_) -> "0".

timestamp() ->
    integer_to_list(unix_time()).

connect_collectd(St) ->
    case connect_collectd(St#st.socket_path, St#st.connect_timeout) of
        { ok, Sock } -> { ok, St#st { socket = Sock }};
        Err -> Err
    end.

connect_collectd(SocketPath, ConnectTimeout) ->
    gen_tcp:connect({local, SocketPath}, 0, [{active, false}, {mode, binary}, local], ConnectTimeout).

unix_time() ->
    datetime_to_unix_time(erlang:universaltime()).

datetime_to_unix_time({{_,_,_},{_,_,_}} = DateTime) ->
    calendar:datetime_to_gregorian_seconds(DateTime) - ?UNIX_EPOCH.

get_opt(K, Opts, Default) ->
    exometer_util:get_opt(K, Opts, Default).

check_instance("auto") ->
    get_default_instance();
check_instance(Other) ->
    Other.

check_hostname("auto") ->
    net_adm:localhost();
check_hostname(H) ->
    H.


get_default_instance() ->
    FullName = atom_to_list(node()),
    case  string:rchr(FullName, $@) of
        0 -> FullName;
        Ind -> string:substr(FullName, 1, Ind - 1)
    end.

%% Parse a line returned by collectd.
%% It has the format
parse_reply(Request, Reply, St) ->
    case parse_reply(Reply, []) of
        {0, _} ->
            {ok, St};

        {-1, _} ->
            ?log(error, "Failed to log ~p: ~p~n", [Request, Reply]),
            { error, St };

        {_, _} ->
            ?log(info, "Got unexpected (and ignored) reply for: ~p: ~p~n", [Request, Reply]),
            { unsupported,  St }
        end.



%% Parse the space after the integer at line beginning.
%% The remainder will always have at least a newline.
parse_reply(<< $\s, Rem/binary >>, RetVal) ->
    %% Reverse the list containing the integer (in ascii format),
    %% and trigger the parsing of the remaining
    Text = binary:part(Rem, 0, size(Rem) - 1),

    %% Reverse the retval and convert to integer.
    %% Return together with text.
    { list_to_integer(lists:reverse(RetVal)), Text };

%% Parse the first part of RetVal, which is the integer at the beginning
%% of the line.
parse_reply(<< C:1/integer-unit:8,Rem/binary >>, RetVal) ->
    parse_reply(Rem, [ C | RetVal ]).

get_type(TypeMap, Extra, Name) ->
    Res = exometer_util:report_type(Name, Extra, TypeMap),
    %% Res = get_opt(Name, TypeMap, undefined),
    ?log(debug, "type_map(~p) -> ~p~n", [ Name, Res ]),
    Res.

maybe_reconnect_after(Socket) ->
    %% Close socket if open
    if Socket =/= undefined -> gen_tcp:close(Socket);
        true -> true
    end,
    prepare_reconnect().

prepare_reconnect() ->
    self() ! {exometer_callback, prepare_reconnect}.

reconnect_after(ReconnectInterval) ->
   erlang:send_after(ReconnectInterval, self(), {exometer_callback, reconnect}).

setup_refresh(RefreshInterval, Metric, DataPoint, Extra, Value) ->
    ?log(debug, "Will refresh after ~p~n", [ RefreshInterval ]),
    TRef = erlang:send_after(RefreshInterval, self(),
                             {exometer_callback, refresh_metric,
                              Metric, DataPoint, Extra, Value}),

    ets:insert(exometer_collectd, { ets_key(Metric, DataPoint), TRef}),
    ok.
